File: ios/GuardApp/Core/LocalNetwork/LanHeartbeatMessage.swift
import CryptoKit
import Foundation

struct LanHeartbeatMessage: Codable, Equatable {
    let protocolVersion: Int
    let groupHash: String
    let deviceId: String
    let deviceNameHash: String
    let timestamp: Int64
    let nonce: String
    let mode: GuardMode
    let signature: String
}

final class LanHeartbeatSigner {
    private let protocolVersion = 1
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    func create(
        groupId: String,
        groupSecret: Data,
        deviceId: String,
        deviceName: String,
        mode: GuardMode,
        timestamp: Int64,
        nonce: String = UUID().uuidString
    ) -> LanHeartbeatMessage {
        let groupHash = shortHash(groupId)
        let deviceNameHash = shortHash(deviceName)
        let payload = signingPayload(
            protocolVersion: protocolVersion,
            groupHash: groupHash,
            deviceId: deviceId,
            deviceNameHash: deviceNameHash,
            timestamp: timestamp,
            nonce: nonce,
            mode: mode
        )
        return LanHeartbeatMessage(
            protocolVersion: protocolVersion,
            groupHash: groupHash,
            deviceId: deviceId,
            deviceNameHash: deviceNameHash,
            timestamp: timestamp,
            nonce: nonce,
            mode: mode,
            signature: hmac(payload, secret: groupSecret)
        )
    }

    func encode(_ message: LanHeartbeatMessage) throws -> Data {
        try encoder.encode(message)
    }

    func decode(_ data: Data) throws -> LanHeartbeatMessage {
        try decoder.decode(LanHeartbeatMessage.self, from: data)
    }

    func verify(_ message: LanHeartbeatMessage, groupSecret: Data, now: Int64, allowedSkewSeconds: Int64 = 30) -> Bool {
        guard message.protocolVersion == protocolVersion else { return false }
        guard abs(now - message.timestamp) <= allowedSkewSeconds else { return false }
        let payload = signingPayload(
            protocolVersion: message.protocolVersion,
            groupHash: message.groupHash,
            deviceId: message.deviceId,
            deviceNameHash: message.deviceNameHash,
            timestamp: message.timestamp,
            nonce: message.nonce,
            mode: message.mode
        )
        return hmac(payload, secret: groupSecret) == message.signature
    }

    private func signingPayload(
        protocolVersion: Int,
        groupHash: String,
        deviceId: String,
        deviceNameHash: String,
        timestamp: Int64,
        nonce: String,
        mode: GuardMode
    ) -> String {
        [
            String(protocolVersion),
            groupHash,
            deviceId,
            deviceNameHash,
            String(timestamp),
            nonce,
            mode.rawValue
        ].joined(separator: "|")
    }

    private func hmac(_ payload: String, secret: Data) -> String {
        let key = SymmetricKey(data: secret)
        let signature = HMAC<SHA256>.authenticationCode(for: Data(payload.utf8), using: key)
        return Data(signature).map { String(format: "%02x", $0) }.joined()
    }

    private func shortHash(_ value: String) -> String {
        let digest = SHA256.hash(data: Data(value.utf8))
        return Data(digest).prefix(8).map { String(format: "%02x", $0) }.joined()
    }
}

File: android/app/src/main/java/com/deviceguard/core/localnetwork/LanHeartbeatMessage.kt
package com.deviceguard.core.localnetwork

import com.deviceguard.core.guardengine.GuardMode
import org.json.JSONObject
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.math.abs

data class LanHeartbeatMessage(
    val protocolVersion: Int,
    val groupHash: String,
    val deviceId: String,
    val deviceNameHash: String,
    val timestamp: Long,
    val nonce: String,
    val mode: GuardMode,
    val signature: String
)

class LanHeartbeatSigner {
    fun create(
        groupId: String,
        groupSecret: ByteArray,
        deviceId: String,
        deviceName: String,
        mode: GuardMode,
        timestamp: Long,
        nonce: String
    ): LanHeartbeatMessage {
        val groupHash = shortHash(groupId)
        val deviceNameHash = shortHash(deviceName)
        val payload = signingPayload(
            protocolVersion = PROTOCOL_VERSION,
            groupHash = groupHash,
            deviceId = deviceId,
            deviceNameHash = deviceNameHash,
            timestamp = timestamp,
            nonce = nonce,
            mode = mode
        )
        return LanHeartbeatMessage(
            protocolVersion = PROTOCOL_VERSION,
            groupHash = groupHash,
            deviceId = deviceId,
            deviceNameHash = deviceNameHash,
            timestamp = timestamp,
            nonce = nonce,
            mode = mode,
            signature = hmacHex(groupSecret, payload)
        )
    }

    fun encode(message: LanHeartbeatMessage): String {
        return JSONObject()
            .put("protocolVersion", message.protocolVersion)
            .put("groupHash", message.groupHash)
            .put("deviceId", message.deviceId)
            .put("deviceNameHash", message.deviceNameHash)
            .put("timestamp", message.timestamp)
            .put("nonce", message.nonce)
            .put("mode", message.mode.wireValue)
            .put("signature", message.signature)
            .toString()
    }

    fun decode(payload: String): LanHeartbeatMessage {
        val json = JSONObject(payload)
        return LanHeartbeatMessage(
            protocolVersion = json.getInt("protocolVersion"),
            groupHash = json.getString("groupHash"),
            deviceId = json.getString("deviceId"),
            deviceNameHash = json.getString("deviceNameHash"),
            timestamp = json.getLong("timestamp"),
            nonce = json.getString("nonce"),
            mode = GuardMode.fromWireValue(json.getString("mode")),
            signature = json.getString("signature")
        )
    }

    fun verify(
        message: LanHeartbeatMessage,
        groupSecret: ByteArray,
        now: Long,
        allowedSkewSeconds: Long = 30
    ): Boolean {
        if (message.protocolVersion != PROTOCOL_VERSION) return false
        if (abs(now - message.timestamp) > allowedSkewSeconds) return false
        val payload = signingPayload(
            protocolVersion = message.protocolVersion,
            groupHash = message.groupHash,
            deviceId = message.deviceId,
            deviceNameHash = message.deviceNameHash,
            timestamp = message.timestamp,
            nonce = message.nonce,
            mode = message.mode
        )
        return hmacHex(groupSecret, payload) == message.signature
    }

    private fun signingPayload(
        protocolVersion: Int,
        groupHash: String,
        deviceId: String,
        deviceNameHash: String,
        timestamp: Long,
        nonce: String,
        mode: GuardMode
    ): String = listOf(
        protocolVersion.toString(),
        groupHash,
        deviceId,
        deviceNameHash,
        timestamp.toString(),
        nonce,
        mode.wireValue
    ).joinToString("|")

    private fun shortHash(value: String): String {
        return MessageDigest.getInstance("SHA-256")
            .digest(value.toByteArray(Charsets.UTF_8))
            .take(8)
            .joinToString("") { "%02x".format(it) }
    }

    private fun hmacHex(secret: ByteArray, payload: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        return mac.doFinal(payload.toByteArray(Charsets.UTF_8)).joinToString("") { "%02x".format(it) }
    }

    private companion object {
        const val PROTOCOL_VERSION = 1
    }
}
